Add on_block_break as an activation trigger, adds cancellability for it

Oblivire opened the pull request:
Adds an activation trigger for when the player breaks a block, and allows for the event to be cancelled in order to either prevent the block from breaking entirely, or prevent it from dropping loot.

The primary use case of this would be to combine it with mcfunctions (though I'm not quite sure how yet) to conditionally be unable to break blocks, silk touch blocks, or just provide a temporary haste when mining certain blocks.

Currently only minorly tested because I'm not very good at making the datapacks - it does appear to at least trigger as intended, though as mentioned I'm not sure how to replace the target block in the action it triggers.

Psither commented:
What is the use case for this cancel variable?

Psither commented:
Why do you need to change every single ability definition to do whatever it is you are planning?

Oblivire commented:
Apologies for the delay in reply:
The use case of the cancel variable is so that after filtering and testing the predicate of the abilities, but before actually running them, the event can see if the event should be cancelled or not
In the case of the on block break event, this prevents the block from being broken after the abilites have ran, so they could break the block and give the item to the player (silk touch) or do nothing (indestructible)

As for why I need to change every single ability definition... I'm honestly not sure that I do?  I was trying to keep it in the same section as "Can be manually disabled", since I was under the impression the other activation events could be cancelled, too
But since the 'post' part of the on target hit and on self hit triggers is being used instead of 'pre' those events can't be cancelled
And on death already has cancellation handling

I also wasn't able to put it next to the condition when I tried to, but I can try again.  In that case, maybe instead of "Cancel" it would be "Silk touch/indestructible/none", which could be handled inside of the event instead of inside of the ability

SiverDX commented:
`cancel` is not needed for every ability
you can put a flag in the on_block_break trigger itself if you want
though I'm a bit doubtful about the need of this

also this seems a bit weird
https://github.com/DragonSurvivalTeam/DragonSurvival/blob/daaf70d9314b2d230cf9bf725c2264ed604364b9/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/ability/activation/trigger/OnBlockBreak.java#L41-L49

why was this added?
https://github.com/DragonSurvivalTeam/DragonSurvival/blob/daaf70d9314b2d230cf9bf725c2264ed604364b9/src/main/java/by/dragonsurvivalteam/dragonsurvival/common/codecs/Condition.java#L130-L137

Oblivire commented:
woops I forgot to remove that second one...
I had added it because I was initially planning to add an eat event hook, before finding that was covered already by the diet.
Sorry!  Will remove it.

As for the first one, my idea was so that the cancellable abilities would tick first and if they do cancel the event, then the other abilities wouldn't trigger.  
The idea being that you could say, have something that triggers on breaking one type of block that silk touches it, and then a different ability that triggers on every block.

Also thank you for the feedback!  Will implement once I get the opportunity to

SiverDX commented:
but why is more than 1 cancelling ability allowed to run
what is the difference to a non-cancelling ability in terms of logic

I don't see a reason why'd you separate them - either only 1 runs and stops all others or all run
the cancel is for the event, not for the interactions between other abilities to my understanding

so if I was to do this, I'd do one loop and just memorize whether one was cancelling and after all are processed potentially cancel the event
because having 1 cancel any other or all would be inconsistent behaviour imo
(you don't know which one will be chosen and you may have one that should always run but shouldn't cancel the event)

apart from the logic there is also the formatting
- usage of `@NotNull` in a local variable
- usage of `snake_case`
- instead of checking whethere the boolean is present you can just do `getOrDefault` and return `List.of`
- (or store it in a variable and check for `null`)

Oblivire commented:
That's fair, I do agree that it would be inconsistent, the more that I look at it.
My initial reasoning is that then the datapack/species creator could have them have the cancelling abilities turn the non-cancelling abilities 'off', but it makes more sense that if you want to do that, you'd just change the predicate

Also apologies for the formatting and poor code, I'm rather unfamiliar with java's syntax.

Hopefully this version's a bit better?  I did include the silk touch part of it, since I'm unsure if that can be handled by an mcfunction or not
Admittedly if it can be, then I wouldn't need to split the silk and indestructible options, it would just be a bool again

SiverDX commented:
in which case would the block need to be indestructible? what kind of ability would that be?
i initially assumed it was just for silk touch behaviour

---

if you do variants you should be using an enum, example:
https://github.com/DragonSurvivalTeam/DragonSurvival/blob/eef89326e1564039b86299755d19b02dd52b22d8/src/main/java/by/dragonsurvivalteam/dragonsurvival/common/codecs/BlockVision.java#L183-L197

---

you can collect the existing variants while ticking the abilities, no need to loop through them multiple times

---

i feel like if you need to implement silk touch explicitly you should call the actual loot table
e.g. through `Block#dropResources` by passing in a proper tool with silk touch?

unless your silk touch is not a "vanilla" silk touch but rather a "drop this exact block"

Oblivire commented:
Thank you!  Implemented using Enum instead

Honestly I'm not too certain of the use-case for the indestructible blocks, I just thought it would make sense when adding silk touch.  
It probably would be more of a penalty than an ability, but would be for say, preventing breaking blocks while in sunlight/rain

Changed to using a silk touch tool instead - it doesn't have to drop the exact block
I think the only reason for getting the exact block would be for getting say, a mob spawner - but a regular player won't be able to place those down without losing the mob anyways, so there's no point

Unfortunately I'm not really sure how to include ability.tick and the enum check in a forEach or map, so I'm using a for loop instead
I also think initializePickaxe could be in a better place, though I'm not sure how to check when SILK_TOUCH has loaded

SiverDX commented:
Been thinking - the on block break trigger can probably stay (though we may want to check the performance impact)

But wouldn't the silk touch part not be a better fit in an ability effect that can grant you the effect of any enchantment?

Oblivire commented:
I agree, though it still needs to cancel if that's the case to prevent double drops - probably will check for that ability effect instead of the enum?

SiverDX commented:
Would it though?

The trigger happens before the event is finished, meaning by the point the loot is collected you'd already have the silk touch enchantment effect

Oblivire commented:
Oh!  I see what you mean now, though I'm not really sure how to implement that?
Doing it for the block break effect would be pretty simple since dropResources accepts a tool that can have an enchantment, but I don't think entities have something similar
It would bring it more in line with how OnDeath cancels, where instead of checking an enum for cancelling, it would compare blockStates before and after the abilities tick, to see if the block has changed

SiverDX commented:
i don't think a harvest effect like that is needed
you also wouldn't need to cancel the block breaking (since the triggers are not used for penalties)
(because the loot that occurs on said block break would have the newly added enchantment taken into account)

you'd need a data attachment which stores the added enchantments
something like https://github.com/DragonSurvivalTeam/DragonSurvival/blob/1.21.1/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/ability/entity_effects/DamageModificationEffect.java
- you have the effect
- the data attachment which holds instances (with a duration)
- you need to hook into the spots where enchantments are checked and check said instances

there would be more things to consider
- is there an event that can be used which does not result in modification of all items?
- (e.g. is `GetEnchantmentLevelEvent` safe?)
- (there also is `EnchantmentHelper#getEnchantmentLevel` with the enchantment and entity whose items are checked - `curios` uses that to increase the looting level e.g., though it's not enough by itself)
- would the instances need an optional equipment slot - if none is specified the enchantment applies to all slots otherwise just to one (e.g. for mending)?

this might be a bit too complex for this pr tbh

Oblivire commented:
I'm not really sure how to do this, honestly, and I agree it's probably outside the scope of this PR

For the purposes of mining a block, it appears to be used like:
Level#destroyBlock (Interestingly called with ItemStack.EMPTY, I guess doing that makes it fall back to whatever the entity has?)
Block#dropResources (same thing I'm currently using)
Block#getDrops (When building the LootParams I think it uses LootParams#getLuck, though unsure where silk touch is handled)
BlockState#getDrops (Seems to be per instance of block, using the provided loot context)

If this was to be implemented, maybe it would be a mixin to override the ItemStack.EMPTY with whatever tool?
Per https://docs.neoforged.net/docs/blocks/#breaking-a-block it might also be possible to change it in BlockDropsEvent?

I feel like this harvest effect + checking for if the block changed makes more sense than what I had before, though - having a BlockBreakEffect with the previous setup probably would have caused an exception

Cancelling by checking for a blockstate difference also will cover for the BlockBreakEffect and BlockConversionEffect - without the cancellation I'm not really sure what would happen

SiverDX commented:
thing is I'm not sure if it's a good idea to implement a halfway solution because once it is replaced with an actual enchantment functionality it would break existing datapacks

---

I noticed that you have a tool in your codec but you don't use it and instead always use a diamond pickaxe

---

generally this new effect seems to overlap too much with https://github.com/DragonSurvivalTeam/DragonSurvival/blob/002b081ce79524933670a542b4ee901cc64650da/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/ability/block_effects/BlockBreakEffect.java

you could maybe extend the block break effect by an optional tool (item stack codec shouldn't need additional enchantment codec fields, since the stack itself should be definable with enchantments)
and if said tool is passed in, then you call the harvest method instead of Level#destroyBlock

Oblivire commented:
Apologies for the delay

That's fair, though adding a tool as an option for BlockBreakEffect would also not be backwards compatible
I think it uses the pickaxe only if the tool isn't defined?  I have "orElse" there
But it doesn't matter since it's being moved

Added the tool to the blockBreakEffect, and made the logic so that if the tool is present it will always drop resources, but if it's present and dropLoot is false, it will also destroy the block afterwards

SiverDX commented:
I don't think such a field is needed on the trigger 

indestructible blocks don't seem to make much sense for abilities and triggers are not used for penalties atm 

harvesting a block without breaking it doesn't seem to make much sense even as an ability

at most I could see a "get more loot" ability but that seems better implemented as a new effect that modifies dropped loot of stuff

Oblivire commented:
Fair, my use case would be to get the dropped resources of a block before converting that block

But it would be better in conversion effect - so I moved it there

And gotcha!  I won't add the indestructible blocks back then

SiverDX commented:
A bit unsure about having this in conversion as well - doesn't seem to quite fit?

Maybe we do actually make an explicit harvest effect that does not modify the block in terms of destroying or converting it explicitly, but rather just acts as if you mined / interacted with the block?

The idea would be for e.g. berry bushes to be properly harvested (i.e. they lose their berries as if they were right-clicked)
And if you specify normal blocks they'd just drop their loot based on the tool you pass in
(For the start it could be kept simple and just call that existing drop resource logic)

And if you want you can then run the destroy or conversion effect separately
(ability effects *should* be executed in the order you define them in the json, at least that's what I'd expect)

@seelderr any input on this

seelderr commented:
I'm still somewhat confused here. Overall I just think an ability effect of this nature is too complex and is going to repeatedly be an issue when trying to maintain it as we continue development, and from a design perspective would be too complex for the player.

Also, the relationship between a tool and an ability seems strange to me. Rather than generating an item to use with the ability, why not just grant effects/conditions that do what the item does? e.g. write up some custom effect that just grants silk touch, or haste effect, etc.  And if we're destroying/transmuting blocks without a tool at all, then why not use the provided effects to just do that instead?

If you want to improve mining of x-y-z block, I think the harvest bonus is sufficient from a gameplay perspective.

Psither commented:
I agree; you should try to use the existing ability effects to produce whatever result you are trying to implement here, and come up with a clear plan for what you want to implement in the least obtrusive way possible before creating a new ability.

If you can't clearly explain what the use case for this new ability is that isn't already covered, I can't approve it.

Oblivire commented:
Use cases:
- Passive that allows you to silk touch specific blocks when you mine them
- Ability to 'yank' materials out of a block, dropping the resource and then converting them, such as magnetizing metal ore out of its respective block
- Making blocks indestructible, though that's more of a penalty and less useful.

The first use case resolves an issue with the Echo Sculk Dragon where mining the block would give you both the block and the exp it drops from being broken - This could cause infinite exp farming, which is why I want to fix it.
The BlockBreak event needs to be canceled to prevent the exp from dropping.

I could add a new Mob Effect to the utils mod that gives silk touch, but the issue I'm running into is that it would need to be for a specific type of block (sculk) 
The effect could be useful for other people instead of just for sculk blocks, so I felt it would make more sense if it was added to Dragon Survival instead of being standalone

It may be feasible to use a mcfunction + the BlockBreak event to drop the resources instead of adding this harvest mechanic, though I don't know if it's possible to drop resources based on a specific block using that.
To my knowledge there isn't any existing effect in Dragon Survival that can do this other than potentially RunFunction - the BlockBreakEffect doesn't work for this as the silk touch loot is not the same as the loot you get with destroyBlock.

The only two options I can see for the first use case are:
- Create a new ItemEntity of the block, spawn it via. addFreshEntity, destroy the block without drops, then cancel the BlockBreak event (As I did earlier [here](https://github.com/Oblivire/DragonSurvival/blob/ef22f72ecba23aae827a4540234145d865153312/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/ability/activation/trigger/OnBlockBreak.java#L50-L57))
- Create a silk touch tool capable of mining the block and use dropResources with that tool specified to spawn the ItemEntity, destroy the block without drops, then cancel the BlockBreak event (current implementation, suggested by Siver over the first one)

The second option would also cover my second use case by dropping the resources without breaking the block, though that would be for a much less developed idea I have instead of the existing sculk dragon

SiverDX commented:
The way I see it there are two options:

---

Implement an `enchantment_effect` like I described here https://github.com/DragonSurvivalTeam/DragonSurvival/pull/827#issuecomment-2870889900 and give you silk touch for 1 tick
But even then I feel like it could carry over to other blocks you break which you didn't intend the silk touch to apply to

---

Do it with the initial approach of a `harvest_effect`
In my view it would have a codec registry or we'd do it like in the projectile effect where we check different codecs
The following types with their own fields could be present:
- drop_resources with a tool field
  - tool doesn't seem too unnatural for this - that's just how minecraft works for this
- harvest crops with a flag to re plant them
- simulate right click

---

Another option would be something like a `summon_tool` in which, for a duration, it acts like you're holding said tool
This one feels a bit weird because you'd have to decide when to use it

Even if you pick "when no item is held" it would conflict with the claw inventory

---

There would be a predicate to which blocks this apply
This doesn't seem *too* complex in my opinion

Personally I think we could implement the `enchantment_effect` and `harvest_effect` at some point
For the sake of this pr I'd start with a simple iteration of the `harvest_effect` though

Don't even need the codec registry etc. for now
Since we can add it later and use the codec with alternative to parse it backwards-compatible
(And remove that part in a major version switch)

Oblivire commented:
I agree, Siver

The first option was how I was thinking of implementing it when standalone - a dummy MobEffect that onBlockBreak would check for the presence of, along with a tag comparison on the block being broken
But this is basically the same thing as just using a predicate with the dropResources and then the BlockBreakEffect, just more limited.
In theory it's possible with just the onBlockBreak event itself if you have a silk touch potion effect from somewhere, since you could make that as an activation trigger to give a potion effect with a duration of 1.  I think in the initial PR comment I suggested using that for haste, though you could do anything with that

The third option would probably require a lot of new mixins, which I was hoping to avoid.  I'm also not really sure where minecraft checks the tool's enchantment?  It seems to just be a lookup when using dropResources in level#destroyBlock that checks the enchantment of either the main hand tool or the tool passed into the method, giving the 'silk touch' loot table if the enchantment is present.  At least that's the conclusion I reached when I made [this comment]( https://github.com/DragonSurvivalTeam/DragonSurvival/pull/827#issuecomment-2873239483)

It might be possible to do this with a Global Loot Modifier, but I can't really see how, since that's non-destructive and only adds things, to my knowledge.

---

The harvest_effect makes the most sense assuming that the effects apply in the same order every time, if they don't then it would probably need to be part of BlockBreak and BlockConversion.  I'll revert to the commit where I added the BlockHarvestEffect.

Though, one caveat.  I don't think we need the right click part of it?  That appears to be covered by [UseItemOnBlockEffect](https://github.com/DragonSurvivalTeam/DragonSurvival/blob/1.21.1/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/ability/block_effects/UseItemOnBlockEffect.java) already, assuming that allows you to use ItemStack.EMPTY

Oblivire commented:
Also tested and broke some blocks with this structure:
```
{
  "activation": {
    "activation_type": "dragonsurvival:passive",
	"trigger": {
      "trigger_type": "dragonsurvival:on_block_break",
	  "condition": {
        "condition": "minecraft:block_state_property",
        "block": "minecraft:grass_block"
      }
    }
  },
  "actions": [
    {
      "target_selection": {
        "target_type": "dragonsurvival:looking_at",
        "applied_effects": {
          "block_effect": [
            {
              "effect_type": "dragonsurvival:block_harvest",
			  "valid_blocks": {
                "type": "minecraft:matching_blocks",
                "blocks": "minecraft:grass_block"
              },
              "probability": 1,
              "tool": {
                "id": "minecraft:diamond_pickaxe",
                "components": {
                  "minecraft:enchantments": {
                    "minecraft:silk_touch": 1
                  }
                },
                "count": 1
			  }
            },
            {
              "effect_type": "dragonsurvival:block_break",
              "valid_blocks": {
                "type": "minecraft:matching_blocks",
                "blocks": "minecraft:grass_block"
              },
              "probability": 1,
              "drop_loot": false
            }
          ]
        },
        "range": 15
      },
      "trigger_rate": 0
    }
  ],
  "can_be_manually_disabled": true,
  "icon": {
    "texture_entries": [
      {
      "from_level": 0,
      "texture_resource": "dragonsurvival:abilities/echo_sculk/sculk_bloom"
    }
  ]
  }
}
```
It appeared to work properly.
I'm not certain that "looking_at" is the best target_selection for this, but I think it's fine enough for testing purposes

Oblivire commented:
Thank you!
